### DISCUSSION TITLE
`branch set` creates branches if they don't already exist?

### DISCUSSION THREAD

**Discussion by BatmanAoD** [jj-vcs/jj#2555] [Q&A] [answered]
Is this behavior intentional? `jj set --help` doesn't say anything about what happens if a branch doesn't exist, and I assumed that branches could only be created via a synchronization operation (e.g. `jj git fetch`) or with `branch create`.

```
❯ : jj branch set revise-result-structure-r @-  --allow-backwards
warning: Updating multiple branches (2).

❯ : jj status
The working copy is clean
Working copy : ywuxrymu 589e7dac @- revise-result-structure-r | (empty) (no description set)
Parent commit: wwqlmqkn 58ae8633 revise-result-structure | <commit message>
```

The fact that a branch named `@-` was created is especially confusing; it's probably worth forbidding branch names starting with characters that have a special meaning in the revset language, such as `~` or `@`.

**Comment by necauqua** [comment, marked as answer]
> Is this behavior intentional?

Yes, although why `branch create` is a thing is interesting, never used it

> a branch named `@-` was created

Also happened to me a few times, I agree this should probably fail

**Reply by ilyagr** [reply]
I believe `jj branch create` will never move a branch.

`jj branch set` is a bit ambiguous; I was also surprised when I discovered it can create a branch. 

I wonder whether it would feel much natural if we *also* had a third command, `jj branch move`, that never creates a branch. Then, `set` is the same as "move or create".

The same thing could be achieved by adding command-line options to distinguish the cases, but it seems less elegant in this case.

**Reply by yuja** [reply]
I'd rather remove `jj branch create` since it's just a stricter version of `jj branch set`. I've never used `create`.

For moving branches around, some automated command to pull up the closest branch might be useful. Say, `jj branch advance` to move `heads(::@ & branches())`'s branch to `@` or `@-`.

**Reply by ilyagr** [reply]
I like `jj branch move` because it's helpful to avoid typos, especially until we can autocomplete branches. I  had confusing misadventures with `jj set typoe`, since for some reason I always end up having dozens of branches and want to move them in confusing directions.

(Rambling aside: A separate story is that my branches need some sort of version control. I keep having branches `feature`, `featore-old`, `feature-tmp`, ... 😃 Also note the typo, though I don't use the phone keyboard with jj usually.)

If we had to have fewer than three, I'd pick `move` and `create` for orthogonality. I'm OK with `move`and `set` if someone -- you, for instance -- preferred that. I can't think of a good reason to stick with `create` and `set` at the moment.

**Reply by yuja** [reply]
I personally don't mind if `jj branch set` refused to create new branch (as I usually don't create named branches.) I just feel having `set`/`create` is redundant.

> it's helpful to avoid typos

Yeah, I understand that. I occasionally find a repo has "main" and "master" and need to figure out which one is the main branch. (the repo also had a branch named "list" btw.) I think the typo problem can be mitigated if `jj branch set` printed "Created"/"Moved branch <name> blah, blah..."

Regarding the command name, I slightly prefer `set` because `jj branch move -rREV` sounds a bit weird.

**Reply by ilyagr** [reply]
>  I think the typo problem can be mitigated if jj branch set printed "Created"/"Moved branch blah, blah..."

This seems like a good idea regardless of anything else.

> I slightly prefer set because jj branch move -rREV sounds a bit weird.

`set` has a similar problem. If we stick with it, we need to update its docs to make it obvious that it has two functions. It's not obvious from the name.

We could also go with `jj branch move --to REV`. I'd keep `-r` as an alias of `--to` (hidden or not), but we'd use `--to` in the docs.

-----------

It seems like different options are slightly better for different workflows. At the moment, I still mildly prefer replacing `set` with `move`, but I don't have a good sense of how inconvenient that is for workflows other than mine. I'm also OK with the plan of softly deprecating `create` (either with the intent of eventually removing it or of making it a long-term alias for `set`) and changing the docs for `set` + adding a message. If we go with the latter plan, we will see how well it works for me and others; if typos remain a bother, I might advocate for adding `move` later.

**Reply by martinvonz** [reply]
@arxanas (who created the `jj branch create/set` commands), what do you think?

**Reply by arxanas** [reply]
In terms of UX, users are surprised by the ability of `jj branch set` to also create a branch if it doesn't exist — or, generally, the existence and usage of a single command which can both create and/or update a value. It's a very implementation-centric way of thinking, and doesn't necessarily map to mental models, except to the extent that we've gotten used to it as users 🙂. We've seen it cause problems in practice when people create branches via `git branch`/`jj branch set`/etc. due to a typo or already-deleted branch or similar.

I think the best approach from a UX perspective is to add a `--create` flag to `jj branch set` to allow creating the branch (for people who use that workflow), and disallow branch creation in `jj branch set` otherwise. I don't feel strongly about `set` vs `move`, except that `--create` sounds a little more natural on `set` to me.

> and I assumed that branches could only be created via a synchronization operation (e.g. jj git fetch) or with branch create.

Here is the empirical evidence that users do not expect that `jj branch set` can create branches 🙂. I imagine we could quickly find more supporting examples in various Git support forums. We should consider the ability of `jj branch set` to create branches to be more of a power-user workflow.

**Reply by ilyagr** [reply]
@arxanas, is your ideal outcome to just have `jj branch set` for moving, `jj branch set --create` (for moving and creating? Creating only?), and not `jj branch create`? Or do you want all three?

**Reply by ilyagr** [reply]
I had another thought about @yuja 's comment:

>  I slightly prefer set because jj branch move -rREV sounds a bit weird.

In addition to the idea of `--to` from https://github.com/martinvonz/jj/discussions/2555#discussioncomment-7525536, there is also an option of renaming `move` to `update`. `jj branch update -rREV` sounds reasonably natural to me for a command that moves branches but does not create them. Optionally, `move` could be an alias.

Aside: Using `update` instead of `set` (as opposed to `move`) might also be an option, though I'm leaning against it.

**Reply by yuja** [reply]
> the best approach from a UX perspective is to add a `--create` flag

If "create" and "set" are mentally different, I think they should be separate commands. (i.e. just make `jj branch set` refuse to create new branch.)

> In addition to the idea of `--to` from [#2555 (reply in thread)](https://github.com/martinvonz/jj/discussions/2555#discussioncomment-7525536), there is also an option of renaming `move` to `update`.

To me, `move` or `update` sounds it would do non-trivial thing (such as rebasing the existing branch somewhere.) As for `jj branch move --to REV`, I would naturally think there would be `--from` to move branches from one revision to the other revision.

**Reply by ilyagr** [reply]
> If "create" and "set" are mentally different, I think they should be separate commands. (i.e. just make jj branch set refuse to create new branch.)

This seems like it would be good enough AFAICT. I had the impression @yuja didn't like this idea, but I may have misunderstood the issue. If we can all actually be happy with it, great!

In terms of the name, if your preference is to stick with the name `set`, I'm completely fine with that. I suggested `move` because I think of this as "moving branches" but, as you pointed out, it's not as perfect a choice as I thought it was.

**Reply by arxanas** [reply]
> @arxanas, is your ideal outcome to just have jj branch set for moving, jj branch set --create (for moving and creating? Creating only?), and not jj branch create? Or do you want all three?

> If "create" and "set" are mentally different, I think they should be separate commands. (i.e. just make jj branch set refuse to create new branch.)

The best UX is to have `jj branch create` and `jj branch set` both exist, and (without additional options) they shouldn't overlap in functionality. That is, `jj branch create` should not update existing branches, and `jj branch set` should not create branches if they don't already exist.

I additionally meant to propose adding `jj branch set --create` to support workflows of current users who want that behavior, but it's not critical for the UX design itself. I can certainly imagine that some people would like to create/update the branch regardless of itself previous status, but `jj branch set` shouldn't do that by default.
